go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,269 @@
+```
+
+Proposal:
+
+```go
+type QueryInput struct {
+  QueryType string
+  Args []interface{}
+}
+
+HandleQuery(ctx Context, input *QueryInput) (interface{}, error)
+```
+
+Why:
+
+* All other calls convert the payload outside of the interceptor call, this can too
+* No need to pass a handler, that is just part of the last item in the interceptor chain (i.e. when calling next)
+
+### WorkflowOutboundCallsInterceptor
+
+We can leave everything the way it is because, while it doesn't make the interceptor parameters future-proof with struct
+wrappers, it mimics the actual calls that aren't going to have backwards-compatible alterations.
+
+### WorkflowOutboundCallsInterceptor.GetWorkflowInfo
+
+Current:
+
+```go
+GetWorkflowInfo(ctx Context) *WorkflowInfo
+```
+
+Proposal:
+
+```go
+GetInfo(ctx Context) *WorkflowInfo
+```
+
+Why: All other calls match `workflow` package, why not this?
+
+### ActivityInterceptor
+
+Proposal:
+
+```go
+type ActivityInterceptor interface {
+  InterceptActivity(ctx context.Context, next ActivityInboundInterceptor) ActivityInboundInterceptor
+}
+```
+
+Why: Mimics `WorkflowInterceptor` but for activities.
+
+### ActivityInboundInterceptor
+
+Proposal:
+
+```go
+type ActivityInput struct {
+  Args []interface{}
+}
+
+type ActivityInboundInterceptor interface {
+  Init(outbound ActivityOutboundInterceptor) error
+
+  ExecuteActivity(ctx context.Context, input *ActivityInput) (interface{}, error)
+}
+```
+
+Also a `ActivityInboundInterceptorBase` implementation with expected delegation.
+
+Why: Mimics `WorkflowInboundInterceptor` but for activities.
+
+### ActivityOutboundInterceptor
+
+Proposal:
+
+```go
+type ActivityOutboundInterceptor interface {
+  GetInfo(ctx context.Context) ActivityInfo
+  GetLogger(ctx Context) log.Logger
+  GetMetricsScope(ctx Context) tally.Scope
+  RecordHeartbeat(ctx context.Context, details ...interface{})
+  HasHeartbeatDetails(ctx context.Context) bool
+  GetHeartbeatDetails(ctx context.Context, d ...interface{}) error
+  GetWorkerStopChannel(ctx context.Context) <-chan struct{}
+}
+```
+
+Also a `ActivityOutboundInterceptorBase` implementation with expected delegation.
+
+Why:
+
+* Mimics `WorkflowOutboundInterceptor` but for activities.
+
+For discussion: `GetInfo` returns expensive copies because `activity.GetInfo` does. Is this ok? This is inconsistent
+with workflow info which is a pointer.
+
+## Interceptor Possible Updates
+
+### Forced Base Embedding
+
+Proposal: Force the base struct to be embedded in all interface implementations. This can be done via a private
+interface method that is only implemented by the base struct.
+
+Why: In the newer gRPC service interfaces this is done to allow interface additions to always be forward compatible. See
+[this](https://github.com/grpc/grpc-go/blob/03ca7b7d00cada2ff8a3ea7348fe6c3a2b2ee4fb/examples/helloworld/helloworld/helloworld_grpc.pb.go#L52)
+as an example.
+
+### WorkerInterceptor
```

Comment:
Do we want this?

Reply:
This seems reasonable to me

Reply:
Agreed, and my tracing interceptors like that activity and workflow interception is combined into a single place like Java.